app/views/questionnaire.py

```diff
@@ -69,50 +72,112 @@ def save_questionnaire_store(response):
-    q_manager = get_questionnaire_manager(g.schema, g.schema_json)
 
-    this_location = Location(group_id, group_instance, block_id)
+    current_location = Location(group_id, group_instance, block_id)
+
+    valid_location = current_location in path_finder.get_routing_path(group_id, group_instance)
+
+    block = _render_schema(current_location)
+
+    error_messages = SchemaHelper.get_messages(g.schema_json)
+
     if 'action[save_sign_out]' in request.form:
-        return _save_sign_out(collection_id, eq_id, form_type, q_manager, this_location)
+        form, _ = post_form_for_location(block, current_location, get_answer_store(current_user), request.form, error_messages, disable_mandatory=True)
+        return _save_sign_out(collection_id, eq_id, form_type, current_location, form)
+
+    form, _ = post_form_for_location(block, current_location, get_answer_store(current_user), request.form, error_messages)
 
-    valid_location = this_location in path_finder.get_routing_path(group_id, group_instance)
-    valid_data = q_manager.validate(this_location, request.form)
+    content = {
+        'form': form,
+        'block': block,
+    }
 
-    if not valid_location or not valid_data:
-        current_location = Location(group_id, group_instance, block_id)
-        _render_schema(current_location)
-        return _build_template(current_location, q_manager.block_state, template='questionnaire')
+    if not valid_location or not form.validate():
+        return _build_template(current_location, content, template='questionnaire')
     else:
-        q_manager.update_questionnaire_store(this_location)
+        update_questionnaire_store_with_form_data(current_location, form.data)
 
-    next_location = path_finder.get_next_location(current_location=this_location)
+    next_location = path_finder.get_next_location(current_location=current_location)
 
     if next_location is None:
         raise NotFound
 
     metadata = get_metadata(current_user)
+
+    return redirect(next_location.url(metadata))
+
+
+@questionnaire_blueprint.route('<group_id>/0/household-composition', methods=["POST"])
+@login_required
+def post_household_composition(eq_id, form_type, collection_id, group_id):
+    path_finder = PathFinder(g.schema_json, get_answer_store(current_user), get_metadata(current_user))
+    answer_store = get_answer_store(current_user)
+
+    current_location = Location(group_id, 0, 'household-composition')
+    block = _render_schema(current_location)
+
+    error_messages = SchemaHelper.get_messages(g.schema_json)
+
+    if 'action[save_continue]' in request.form:
+        _remove_repeating_on_household_answers(answer_store, group_id)
+
+    form, _ = post_form_for_location(block, current_location, answer_store, request.form, error_messages)
+
+    if 'action[add_answer]' in request.form:
+        form, _ = post_form_for_location(block, current_location, answer_store, request.form, error_messages, disable_mandatory=True)
```

Comment:
It doesn't 'just' set a flag, it reconstructs the whole form with mandatory field errors ignored. Once the form is created, it can't be altered. I can't see a way of getting round this I'm afraid, given the default path depends on a non-mandatory form being created.